clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java

```diff
@@ -1497,4 +1505,11 @@ public void handleResponse(AbstractResponse response) {
             }
         }
     }
+
+    private boolean isFatalException(Errors error) {
+        return error == Errors.TRANSACTIONAL_ID_AUTHORIZATION_FAILED
+                   || error == Errors.INVALID_PRODUCER_EPOCH
+                   || error == Errors.UNSUPPORTED_FOR_MESSAGE_FORMAT
+                   || error == Errors.FENCED_INSTANCE_ID;
```

Comment:
I still don't think this should be a fatal error _for the producer_. As long as we can still abort the transaction, it should be an abortable error. It's similar to the handling of `GROUP_AUTHORIZATION_FAILED`.

Reply:
I'm not strong about this either, as consumer should always be the reliable source for fencing, while producer's exception handling could be simplified.

Reply:
Should we make fenced_instance_id non-fatal as well?

Reply:
Just addressed this @guozhangwang

Reply:
Wait, how did we end up back here? I thought we agreed this should not be fatal for the producer? I think it should have a separate branch above, similar to the handling of `GROUP_AUTHORIZATION_FAILED`.

Reply:
Ugh, my b